### PR TITLE
[ntuple] Add chain support for RDF data source

### PR DESCRIPTION
**Pull request by jblomer**
Adds a new constructor to the RNTuple RDF data source that takes an ntuple name and a list of files. The data source then treats these files as a chain.

Slightly simplifies the prototype construction in the RNTuple data source: instead of constructing full column reader prototypes, only the corresponding field prototypes are constructed.

Multi-threaded work scheduling is as follows:

- As long as there are more files than slots: every slots get a file
- For the tail: slots get distributed over the remaining files and align on cluster boundaries

Some optimizations for follow-up PRs:

- Load the next ranges in the background. In particular that would open the next batch of files in the backgound
- Clone page sources without reopening the file
- The next set of ranges currently gives up to number of slots ranges. However, it would make more sense to use resource constraints such as available memory or number of open file descriptors as a limit and let slots work on more ranges at a time.

### CODE REVIEW

**Comment by jblomer** [comment]
There is still the non-trivial question of how to distribute work to slots. So while this change makes it possible to process RNTuple chains with RDF, it's rather sub optimal because parallelization is only within a single file and it is not respecting cluster boundaries.

**Comment by eguiraud** [comment]
About work distribution, IIUC currently each sub-rntuple in a chain is processed sequentially, with parallelism possible within the processing of each sub-rntuple.

Unfortunately with the current RDataSource API we cannot pull a trick like distRDF does, distributing ranges as file %'s and figuring out exact event-level splitting later. One pass over the full chain might be necessary, at least for now.

File: tree/dataframe/inc/ROOT/RNTupleDS.hxx [hunk 1]
```diff
@@ -44,19 +45,53 @@ class RNTupleColumnReader;
...
-   unsigned fNSlots = 0;
-   bool fHasSeenAllRanges = false;
+   /// List of column readers returned by GetColumnReaders() organized by slot. Used to reconnect readers
+   /// to new page sources when the files in the chain change.
+   std::vector<std::vector<Internal::RNTupleColumnReader *>> fActiveColumnReaders;
+
+   unsigned int fNSlots = 0;
+   ULong64_t fSeenEntries = 0;              ///< The number of entries so far returned by GetEntryRanges()
+   std::vector<REntryRange> fCurrentRanges; ///< Basis for the ranges returned by the last GetEntryRanges() call
+   std::vector<REntryRange> fNextRanges;    ///< Basis for the ranges returned by the next GetEntryRanges() call
```
Comment (enirolf) [inline tree/dataframe/inc/ROOT/RNTupleDS.hxx:90]:
```suggestion
   std::vector<REntryRange> fNextRanges;    ///< Basis for the ranges populated by the PrepareNextRanges() call
```

**Review by enirolf — CHANGES_REQUESTED** [review]
Many thanks! I've added a few comments and clarifying questions. Maybe a general question: there are a few `R__ASSERT`s. Is there any way these could be hit by users, and should those be turned into exceptions instead?

File: tree/dataframe/src/RNTupleDS.cxx [hunk 2]
```diff
@@ -114,38 +114,70 @@ class RNTupleColumnReader : public ROOT::Detail::RDF::RColumnReaderBase {
...
+   void Connect(RPageSource &source, ULong64_t entryOffset)
    {
+      R__ASSERT(fLastEntry == -1);
+      fEntryOffset = entryOffset;
+
+      // Create a new, real field from the prototype and set its field ID in the context of the given page source
+      fField = fProtoField->Clone(fProtoField->GetName());
+      {
+         auto descGuard = source.GetSharedDescriptorGuard();
+         fField->SetOnDiskId(
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/dataframe/src/RNTupleDS.cxx:142]: Two fields (`RFieldBase` objects) created for RDataFrame can end up having the same field ID if they represent the same underlying physical field in different ways. For instance, for a `vector<float>` field, the field as `vector<float>` has the same ID than the field that represents the size of the vector.

On disk, all fields have their own, unique ID.

**Review by vepadulano — APPROVED** [review]
Thank you so much for this! I have a couple of comments regarding parts of the logic, we can iterate over these also in a separate PR, so feel free to merge.

**Review by enirolf — APPROVED** [review]

**Pull request merged**